base/refpointer.jl

```diff
@@ -0,0 +1,121 @@
+### General Methods for Ref{T} type
+
+Base.eltype{T}(x::Type{Ref{T}}) = T
+Base.convert{T}(::Type{Ref{T}}, x::Ref{T}) = x
+
+# create Ref objects for general object conversion
+@inline Base.cconvert_gcroot{T}(::Type{Ref{T}}, x) = convert(Ref{T}, x)
+@inline Base.cconvert{T}(::Type{Ref{T}}, x) = convert(Ptr{T}, x)
+
+### Methods for a Ref object that can store a single value
+
+type RefValue{T} <: Ref{T}
+    x::T
+    RefValue() = new()
+    RefValue(x) = new(x)
+end
+Base.convert{T}(::Type{Ref{T}}, x) = RefValue{T}(x)
+Base.call{T}(::Type{Ref{T}}) = RefValue{T}()
+
+Ref(x::Ref) = x
+Ref{T}(x::T) = RefValue{T}(x)
+Ref{T}(x::Ptr{T}, i::Integer=1) = x + (i-1)*Core.sizeof(T)
+Ref(x, i::Integer) = (i != 1 && error("Object only has one element"); Ref(x))
+
+@inline function Base.convert{T}(P::Type{Ptr{T}}, b::RefValue{T})
+    if isbits(T)
+        return convert(P, data_pointer_from_objref(b))
+    else
+        return convert(P, data_pointer_from_objref(b.x))
```

Comment:
I realize this does the correct thing for ccall, but it's a confusing semantic for a wrapper object to have: sometimes the wrapper pretends not to be there. I would prefer one of these:

1. Only allow constructing these for isbits types.
1. If `x` isn't a bits type just return it and don't wrap it.
1. Rename these types, and have `Ref` be just a simple assignable cell with the obvious behavior.